contracts/MultiPartyEscrow.sol

```diff
@@ -0,0 +1,272 @@
+pragma solidity ^0.4.24;
+
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+
+contract MultiPartyEscrow {
+    
+    //it seems we don't need SafeMath
+   //using SafeMath for uint256;
+    
+
+    //TODO: we could use uint64 for replica_id and nonce (it could be cheaper to store but more expensive to operate with)
+
+    //the full ID of "atomic" payment channel = "[this, channel_id, nonce]"
+    struct PaymentChannel {
+        address sender;      // The account sending payments.
+        address recipient;    // The account receiving the payments.
+        uint256 replica_id;  // id of particular service replica
+        uint256 value;       // Total amount of tokens deposited to the channel. 
+        uint256 nonce;       // "nonce" of the channel (by changing nonce we effectivly close the old channel ([this, channel_id, old_nonce])
+                             //  and open the new channel [this, channel_id, new_nonce])
+                             //!!! nonce also prevents race conditon between channel_claim and channel_extend_and_add_funds 
+        uint256 expiration;  // Timeout in case the recipient never closes.
+    }
+
+
+    mapping (uint256 => PaymentChannel) public channels;
+    mapping (address => uint256)        public balances; //tokens which have been deposit but haven't been escrowed in the channels
+   
+    //already used messages for open_channel_by_recipient in order to prevent replay attack
+    mapping (bytes32 => bool) public used_messages; 
+
+    uint256 public next_channel_id; //id of the next channel (and size of channels)
+ 
+    ERC20 public token; // Address of token contract
+    
+    //TODO: optimize events. Do we need more (or less) events?
```

Comment:
I do think logging is super useful, especially for light clients. I would add also the `close` event.

Reply:
I'm totally agree with you... Events are cheap and we can have more of them... But I will add them later to not spend time now...

Reply:
@astroseger, please open an issue on adding events then. And I would propose removing code which is commented out.

Reply:
I mean, I would use those commented events, LGTM. 

I will open an issue about editing/adding more events.